### PR TITLE
Broken `toArray` implementations in `BuildReferenceMapAdapter`

### PR DESCRIPTION
**Pull request by jglick**
While working on #7998 in some earlier iterations I had stumbled across errors in the collection types used in `BuildReferenceMapAdapter` which I presume were never tested before and are not used in production code.

### Testing done

Without patches, the test fails with the following stack traces:

```
java.lang.StackOverflowError
	at java.base/java.util.ArrayList.<init>(ArrayList.java:179)
	at jenkins.model.lazy.BuildReferenceMapAdapter$SetAdapter.toArray(BuildReferenceMapAdapter.java:312)
	at java.base/java.util.ArrayList.<init>(ArrayList.java:179)
	at …
java.lang.StackOverflowError
	at jenkins.model.lazy.BuildReferenceMapAdapter$CollectionAdapter.toArray(BuildReferenceMapAdapter.java:195)
	at java.base/java.util.ArrayList.<init>(ArrayList.java:179)
	at jenkins.model.lazy.BuildReferenceMapAdapter$CollectionAdapter.toArray(BuildReferenceMapAdapter.java:195)
	at …
java.lang.ArrayIndexOutOfBoundsException: Index 0 out of bounds for length 0
	at jenkins.model.lazy.BuildReferenceMapAdapter$SetAdapter.toArray(BuildReferenceMapAdapter.java:333)
	at jenkins.model.lazy.LazyLoadRunMapEntrySet.toArray(LazyLoadRunMapEntrySet.java:109)
	at java.base/java.util.Collection.toArray(Collection.java:381)
	at jenkins.model.lazy.AbstractLazyLoadRunMapTest.entrySetToArray(AbstractLazyLoadRunMapTest.java:396)
java.lang.ArrayIndexOutOfBoundsException: Index 0 out of bounds for length 0
	at jenkins.model.lazy.BuildReferenceMapAdapter$CollectionAdapter.toArray(BuildReferenceMapAdapter.java:213)
	at java.base/java.util.Collection.toArray(Collection.java:381)
	at java.base/java.util.Collections$UnmodifiableCollection.toArray(Collections.java:1039)
	at jenkins.model.lazy.AbstractLazyLoadRunMapTest.entrySetToArray(AbstractLazyLoadRunMapTest.java:401)
```

### Proposed changelog entries

- Fixing some `toArray` implementations in probably unused code.

### Proposed upgrade guidelines

N/A

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7999"></a>

### CODE REVIEW

**Review by Vlatombe — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]
/label ready-for-merge

---

This PR is now ready for merge. We will merge it after ~24 hours if there is no negative feedback.
Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.
Thanks!

**Pull request merged**
